nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+                        )),
+                        error: None,
+                        instance_updated: None,
+                    };
+                }
+                Err(ClientError::CommunicationError(e)) => {
+                    // TODO(eliza): eventually, we may want to transition the
+                    // instance to the `Failed` state if the sled-agent has been
+                    // unreachable for a while. We may also want to take other
+                    // corrective actions or alert an operator in this case.
+                    slog::info!(opctx.log, "sled agent is unreachable"; "error" => ?e);
+                    return CheckResult {
+                        target,
+                        check_failure: Some(CheckFailure::SledAgentUnreachable),
+                        error: None,
+                        instance_updated: None,
+                    };
+                }
+                Err(e) => {
+                    slog::warn!(
+                        opctx.log,
+                        "error checking up on instance";
+                        "error" => ?e,
+                        "status" => ?e.status(),
+                    );
+                    return CheckResult {
+                        target,
+                        check_failure: None,
+                        error: Some(CheckError::ClientError),
+                        instance_updated: None,
+                    };
+                }
+            };
+
+            slog::debug!(opctx.log, "updating instance state: {state:?}");
+            let update_result = crate::app::instance::notify_instance_updated(
+                &datastore,
+                &resolver,
+                &opctx,
+                &opctx,
+                &opctx.log,
+                &instance_id,
+                &state.into(),
+            )
+            .await
+            .map_err(|_| CheckError::UpdateFailed)
+            .and_then(|updated| updated.ok_or(CheckError::InstanceNotFound));
+            match update_result {
+                Ok(updated) => {
+                    slog::debug!(opctx.log, "update successful"; "instance_updated" => updated.instance_updated, "vmm_updated" => updated.vmm_updated);
+                    CheckResult {
+                        target,
+                        instance_updated: Some(updated),
+                        check_failure: None,
+                        error: None,
+                    }
+                }
+                Err(e) => {
+                    slog::warn!(opctx.log, "error updating instance"; "error" => ?e);
+                    CheckResult {
+                        target,
+                        instance_updated: None,
+                        check_failure: None,
+                        error: Some(e),
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(
+    Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, oximeter::Target,
+)]
+struct InstanceTarget {
```

Comment:
I would probably eschew the use of "target" in the name here, it's mostly `oximeter`-internal jargon. Something like the `VirtualMachine` struct we use in Propolis might be good. The exact same name with different fields will not raise conflict errors, so long as the _full timeseries name_ (target name + metric name) has a unique and consistent set of fields. So don't name any metric here `VcpuUsage` and you should be good :)

In any case, we should definitely add the project and silo IDs as fields to these timeseries. I might also recommend adding the sled's model / revision / serial as well, if they're easily accessible at the point we build  type.

Reply:
Ah, I didn't realize the name of the struct was reflected in the generated metrics, thanks! I'll change it.

Reply:
Out of curiosity, is it possible for other fields like the project/silo/sled serial/etc to be hydrated from the sled and instance UUIDs when the metrics are actually consumed? Or do we need to record everything we might ever care about now?

Reply:
>  I didn't realize the name of the struct was reflected in the generated metrics

I thought I was being clever when I wrote these macros, but this has been a source of confusion for a long time! I am planning some work in the near future that's mostly to support _updates_ to these schemas, and I will be making the name of these things a very explicit part of that.

> Out of curiosity, is it possible for other fields like the project/silo/sled serial/etc to be hydrated from the sled and instance UUIDs when the metrics are actually consumed?

I don't quite follow this. Could you expand on your question? Do you mean something like, looking up the project ID elsewhere at the time we query the metrics?

> Or do we need to record everything we might ever care about now?

Without any mechanisms for updating these at the moment, I would err on the side of more is better. These are quite cheap to store, we only have one copy of them in the database per _timeseries_, not per sample. (Though they are emitted in each sample.)

Reply:
> > Out of curiosity, is it possible for other fields like the project/silo/sled serial/etc to be hydrated from the sled and instance UUIDs when the metrics are actually consumed?
> 
> I don't quite follow this. Could you expand on your question? Do you mean something like, looking up the project ID elsewhere at the time we query the metrics?

Yeah, that's what I was wondering about. My prior experience with metrics systems has mostly been with Prometheus, where it's possible for a query to join additional label fields with a timeseries, so the set of labels added to a particular timeseries need only be enough to uniquely identify that timeseries, and other labels can be hydrated at display time. On the other hand, the number of labels on a timeseries as it's recorded was expensive enough that we generally tried to avoid adding labels that could be joined in a query based on a unique key. If the constraints for Clickhouse are different, that's good to know!

Reply:
> In any case, we should definitely add the project and silo IDs as fields to these timeseries. I might also recommend adding the sled's model / revision / serial as well, if they're easily accessible at the point we build type.

Currently, the CRDB table I'm querying to get the list of instances/sled-agents to check (the `sled_instance` table) has silo and project _names_, but it doesn't have their UUIDs. Do you think it's okay to use the name, or should I change the DB query to do more joins to get the UUIDs instead?

Reply:
> it's possible for a query to join additional label fields with a timeseries

We current support inner joins in OxQL, but not outer joins, which you'd need to combine disparate fields from two timeseries. It shouldn't be too much trouble to do that, if we decide we need it. I can appreciate the desire for a compact representation, but I have consistently found that ClickHouse is able to handle enormous amounts of data. We deduplicate the fields of each timeseries on disk, so you have one copy of them for each timeseries. ClickHouse will further compress them, so the fields themselves are extremely cheap. I would err on the side of attaching all the fields you want, to the timeseries where you'd like them.

> Currently, the CRDB table I'm querying to get the list of instances/sled-agents to check (the sled_instance table) has silo and project names, but it doesn't have their UUIDs. Do you think it's okay to use the name, or should I change the DB query to do more joins to get the UUIDs instead?

I would recommend paying the cost of the extra lookup to get the UUIDs, at some point when you're fetching the data. The names can change at any time in response to a user API request, and I think we'd prefer to be unaware or unencumbered by that in the timeseries system. It is not clear how you'd follow an instance that's renamed, for example, since CRDB stores only the _current_ name, not its history.

Reply:
Okay, that makes sense, thank you! As an aside, though, it occurs to me that, because we can't do outer joins, there isn't presently a way for OxQL to turn a project/silo UUID into a user-understandable name?